biggerquery/workflow.py

```diff
@@ -7,15 +9,164 @@ def __init__(self,
                  schedule_interval=DEFAULT_SCHEDULE_INTERVAL,
                  dt_as_datetime=False,
                  **kwargs):
-        self.definition = definition
+        self.definition = self._parse_definition(definition)
         self.schedule_interval = schedule_interval
         self.dt_as_datetime = dt_as_datetime
         self.kwargs = kwargs
 
     def run(self, runtime):
-        for job in self:
+        for job in self.build_sequential_order():
             job.run(runtime=runtime)
 
-    def __iter__(self):
-        for job in self.definition:
-            yield job
\ No newline at end of file
+    def build_sequential_order(self):
+        return self.definition.sequential_order()
+
+    def call_on_graph_nodes(self, consumer):
+        self.definition.call_on_graph_nodes(consumer)
+
+    def _parse_definition(self, definition):
+        if isinstance(definition, list):
+            return Definition(self._map_to_workflow_jobs(definition))
+        if isinstance(definition, Definition):
+            return definition
+        raise ValueError(f"Invalid argument {definition}")
+
+    @staticmethod
+    def _map_to_workflow_jobs(job_list):
+        return [WorkflowJob(job, i) for i, job in enumerate(job_list)]
+
+
+class WorkflowJob:
+    def __init__(self, job, name):
+        self.job = job
+        self.name = name
+
+    def run(self, runtime):
+        self.job.run(runtime=runtime)
+
+    def __hash__(self):
+        return hash(self.name)
+
+    def __eq__(self, other):
+        return self.name == other.name
+
+    def __str__(self):
+        return "WorkflowJob{job=..., name=%s}" % self.name
+
+
+class Definition:
+    def __init__(self, jobs):
+        self.job_graph = self._build_graph(jobs)
+        self.job_order_resolver = JobOrderResolver(self.job_graph)
+
+    def sequential_order(self):
+        return self.job_order_resolver.find_sequential_run_order()
+
+    def call_on_graph_nodes(self, consumer):
+        return self.job_order_resolver.call_on_graph_nodes(consumer)
+
+    def _build_graph(self, jobs):
+        if isinstance(jobs, list):
+            job_graph = self._convert_list_to_graph(jobs)
+        elif isinstance(jobs, dict):
+            job_graph = jobs
+        else:
+            raise ValueError("Job graph has to be dict or list")
+
+        JobGraphValidator(job_graph).validate()
```

Comment:
Why not just a function?

Reply:
Why as a function :) ? I wanted to clearly separate logic responsible for validation and finding job order running. There are totally different

Reply:
I mean function, not method- validate_job_graph(job_graph). That's not critical, but more pythonic I think.

Reply:
Yep I know but I wanted to clearly separate this block of code and do not move it to another module